jena-arq/src/main/java/org/apache/jena/sparql/core/mem/DatasetGraphInMemory.java

```diff
@@ -125,57 +115,104 @@ public DatasetGraphInMemory(final QuadTable i, final TripleTable t) {
+        startTransaction(readWrite) ;
+        _begin(readWrite) ;
+    }
+
+    private void _begin(ReadWrite readWrite) {
+        withLock(systemLock, () ->{
             quadsIndex().begin(readWrite);
             defaultGraph().begin(readWrite);
-        } finally {
-            commitLock().readLock().unlock();
-        }
+        }) ;
+    }
+    
+    /** Called transaction start code at most once per transaction. */ 
+    private void startTransaction(ReadWrite mode) {
+        writeLock.enterCriticalSection(mode.equals(READ)); // get the dataset write lock, if needed.
+        transactionType(mode);
+        isInTransaction(true);
     }
 
+    /** Called transaction ending code at most once per transaction. */ 
+    private void finishTransaction() {
+        isInTransaction.remove();
+        transactionType.remove();
+        writeLock.leaveCriticalSection();
+    }
+     
     @Override
     public void commit() {
-        if (!isInTransaction()) throw new JenaTransactionException("Tried to commit outside a transaction!");
-        commitLock().writeLock().lock();
-        try {
+        if (!isInTransaction())
+            throw new JenaTransactionException("Tried to commit outside a transaction!");
+        if (transactionType().equals(WRITE))
+            _commit();
+        finishTransaction();
+    }
+
+    private void _commit() {
+        withLock(systemLock, () -> {
             quadsIndex().commit();
             defaultGraph().commit();
-        } finally {
-            commitLock().writeLock().unlock();
-        }
-        isInTransaction.remove();
-        writeLock().leaveCriticalSection();
+            quadsIndex().end();
+            defaultGraph().end();
+        } ) ;
     }
-
+    
     @Override
     public void abort() {
-        if (!isInTransaction()) throw new JenaTransactionException("Tried to abort outside a transaction!");
-        end();
+        if (!isInTransaction()) 
+            throw new JenaTransactionException("Tried to abort outside a transaction!");
+        if (transactionType().equals(WRITE))
+            _abort();
+        finishTransaction();
     }
 
+    private void _abort() {
+        withLock(systemLock, () -> {
+            quadsIndex().abort();
+            defaultGraph().abort();
+            quadsIndex().end();
+            defaultGraph().end();
+        } ) ;
+    }
+    
     @Override
     public void close() {
-        if (isInTransaction()) abort();
+        if (isInTransaction())
+            abort();
     }
 
     @Override
     public void end() {
         if (isInTransaction()) {
-            if (transactionType().equals(WRITE))
-                log.warn("end() called for WRITE transaction without commit or abort having been called");
+            if (transactionType().equals(WRITE)) {
+                log.warn("end() called for WRITE transaction without commit or abort having been called causing a forced abort");
```

Comment:
This message is a tiny bit ambiguous if you don't already understand that `end` calls `_abort`. Maybe just add a sentence break: `"end() called for WRITE transaction without commit or abort having been called. This causes a forced abort!"`